### PR TITLE
Table to store featured project ids

### PR DESCRIPTION
**Pull request by Erin007**
CLOSED in favor of #19510.

### CODE REVIEW

**Comment by Erin007** [comment]
cc: @tanyaparker

File: dashboard/test/factories/featured_projects.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+#  project_id :string(255)
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+
+FactoryGirl.define do
+  factory :featured_project do
+    project_id "MyString"
+  end
+end
```
Comment (aoby) [inline dashboard/test/factories/featured_projects.rb]: This and the below empty test look like leftover `rails generate` scaffolding. Please remove, or fill in with usable code.

File: dashboard/db/migrate/20171204180201_create_featured_projects.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class CreateFeaturedProjects < ActiveRecord::Migration[5.0]
+  def change
+    create_table :featured_projects do |t|
+      t.string :project_id
+
+      t.timestamps
```
Comment (aoby) [inline dashboard/db/migrate/20171204180201_create_featured_projects.rb:6]: Would it be helpful to also record the user id who marked it as a featured project?

**Comment by davidsbailey** [comment]
Hi Erin, unless you plan to add other metadata relating to how a project is featured, it seems like this should just be a new column on storage_apps.

What do the created_at and updated_at fields represent here? if you are trying to capture the date at which a project was featured, you could do it by having the new column be `featured_at`, and having it be null if the project is not featured (very similar to the existing `published_at` column).

If you do want to stick with a separate table, my recommendations would be:
1. use storage_app_id (decrypted, integer, index into storage_apps table) instead of project_id
2. add an index on the storage_app_id column
3. think about whether you need the timestamps or can just join with the storage_apps table to find the timestamps on the underlying project there.

Feel free to solicit a second opinion and/or swing by to chat about this.

File: dashboard/db/migrate/20171204180201_create_featured_projects.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/db/migrate/20171204180201_create_featured_projects.rb:6]: Recording the user id would be a good example of metadata that would justify having a separate table for this.

**Comment by Erin007** [comment]
My thinking was that storage_apps is a big table so adding two additional columns (featured and featured_at), which will more often than not be false/null seemed like storing more information than necessary.  Is there a performance advantage to keeping everything in one table? More broadly, are there rules to consider for relational database design that help inform designs like these? (I don't know that much about it). Also, I know historically why the projects info is stored in Pegasus, but conceptually seems like dynamic projects connected specifically to users ultimately should live in dashboard; maybe starting to store info about featured projects gets us incrementally closer to that ideal? 🤷🏼‍♀️ I'm not 100% committed to a new table versus new columns added to the current table, just trying to understand the pros/cons of each approach.

**Comment by davidsbailey** [comment]
This is a great question Erin. Although I gave you an opinionated initial answer, I can definitely see both sides of it. My thinking is that attributes of the same entity (in this case a project) should go in the same table, unless there's a reason to use a separate table. There is also a rule for performance which says "optimize later". However, in this case, because the table is large (24M rows and growing quickly) and the number of rows with a non-null `featured` fields will be quite small, it could make sense to optimize now. 

@wjordan would be the best person to ask about this. I'd recommend pinging him in the dashboard-dev chat room to see if you can solicit an expert opinion. meanwhile I'll see if I can find any additional expert advice on the internet.

**Comment by aoby** [comment]
The separate table isn't solely a perf optimization (in which case, I'd agree optimize later), but also a different organizational pattern. It's reasonable to discuss the tradeoffs between a more or less normalized structure in the DB schema whenever we add new features. Reorganizing the schema later, especially on a large table, can be cumbersome.

**Comment by aoby** [comment]
I could see this as instead of an attribute on a project (is it marked exemplary or not), this is a separate list of exemplary projects. This way, conceptually as well, the separate table with a foreign key project id might make more sense. 

I don't have a lot of context though, so that might not be a great way to think about it. I'm cool with either approach as well, though leaning toward the separate small table.

**Comment by davidsbailey** [comment]
This article seems to be the best explanation of the decision you're facing from a perf perspective: https://dba.stackexchange.com/questions/18399/database-design-new-table-versus-new-columns , and calls the separate table idea "Vertical Partitioning". My takeaways from it are:
1. optimize later in general, but
2. this case meets all the criteria for when to go with a separate table, most notably because you'll probably never be JOINing across the two tables (except for a single row, and then only for levelbuilder users)

I agree with Andrew that this is also a decision between storing the data in [normalized](https://en.wikipedia.org/wiki/Database_normalization) (separate table) and [denormalized](https://en.wikipedia.org/wiki/Denormalization) (same table) form, however much of the discussion around those formats assumes that "denormalizing" the data will be a performance optimization, which it would not be in this case (because we aren't going to be joining across these two tables very often).

The only downside I see with this approach is a potential proliferation of small tables, but given that there are only a few tables big enough to consider splitting columns off of in this way, I think we are ok.

So, I'm now convinced that a separate table is the best way to go (although a single table would be fine, too).

Thank you for asking good questions about this, Erin!  To repeat/revise my initial comments regarding the two-table approach:
1. use storage_app_id (decrypted, integer, index into storage_apps table) instead of project_id
2. add an index on the storage_app_id column
3. timestamps as you have them seem fine, assuming they represent the time at which the project was featured, although they seem redundant, so if there's a way to just to created_at (but not updated_at) that might be better. @aoby might know the best practice here.

**Comment by Erin007** [comment]
Thanks for the discussion and useful info! I proceed with the separate table and take @davidsbailey's suggestions about storage_app_id rather than project_id, etc. 

@aoby Rails automatically added created_at and updated_at when I generated the new model. Do we usually keep both, neither?

**Comment by aoby** [comment]
@Erin007 from http://edgeguides.rubyonrails.org/active_record_migrations.html:
> The timestamps macro adds two columns, created_at and updated_at. These special columns are automatically managed by Active Record if they exist.

If you only want `created_at`, simply create that directly:
```ruby
t.datetime :created_at
```

Rails (ActiveRecord) will automatically populate that column with the created at time. The only thing that makes these columns "special" is the name, which is independent from how they're created. That `timestamps` macro is just a helpful shorthand for creating both.

**Pull request closed without merging**